test/topology_experimental_raft/test_tablets.py

```diff
@@ -1296,3 +1315,186 @@ async def test_tablet_storage_freeing(manager: ManagerClient):
+    total_replicas = initial_tablets * rf
+    nodes_per_rack: defaultdict[str, set[ServerNum]] = defaultdict(set)
+    for s in servers:
+        nodes_per_rack[s.rack].add(s.server_id)
+    replicas_per_rack = total_replicas / len(nodes_per_rack.keys())
+    result: dict[ServerNum, float] = dict()
+    for rack_servers in nodes_per_rack.values():
+        replicas_per_node = replicas_per_rack / len(rack_servers)
+        for id in rack_servers:
+            result[id] = replicas_per_node
+    #logger.debug(f"get_expected_replicas_per_server: servers={servers} initial_tablets={initial_tablets} rf={rf}: total_replicas={total_replicas} nodes_per_rack={nodes_per_rack}")
+    return result
+
+
+def verify_replicas_per_server(desc: str, expected_replicas_per_server: dict[ServerNum, float], tablet_count: dict[ServerNum, list[int]], initial_tablets: int, rf: int, shards_per_node: int = 2):
+    logger.debug(f"{desc}: expected_replicas_per_server={expected_replicas_per_server} tablet_count={tablet_count}")
+    total = 0
+    for server_id, host_replicas in tablet_count.items():
+        expected_replicas_per_shard = expected_replicas_per_server[server_id] / shards_per_node
+        for i in host_replicas:
+            total += i
+            assert abs(i - expected_replicas_per_shard) <= 1
+    assert total == initial_tablets * rf
+
+
+@pytest.mark.asyncio
+async def test_decommission_rack_basic(manager: ManagerClient):
+    """
+    Test decommissioning of all nodes in a rack
+    when there are enough remaining racks to satisfy
+    the replication factor.
+    """
+    logger.info("Bootstrapping multi-rack cluster")
+    num_racks = 3
+    nodes_per_rack = 2
+    rf = num_racks - 1
+
+    servers = await create_cluster(manager, 1, num_racks, nodes_per_rack)
+    ks, table, initial_tablets, _ = await create_and_populate_table(manager, rf=rf)
+
+    logger.info("Verify tablet replicas distribution")
+    tables = {ks: [table]}
+    expected_replicas_per_server = get_expected_replicas_per_server(list(servers.values()), initial_tablets, rf)
+    tablet_count = await get_tablet_count_per_shard_for_hosts(manager, list(servers.values()), tables)
+    verify_replicas_per_server("Before decommission", expected_replicas_per_server, tablet_count, initial_tablets, rf)
+
+    live_servers: dict[ServerNum, ServerInfo] = dict()
+    dead_servers: dict[ServerNum, ServerInfo] = dict()
+    decommision_rack = f"rack{num_racks}"
+    logger.debug(f"Decommissioning rack={decommision_rack}")
+    for s in servers.values():
+        if s.rack == decommision_rack:
+            logger.debug(f"Decommissioning server={s}")
+            await manager.decommission_node(s.server_id)
+            dead_servers[s.server_id] = s
+        else:
+            live_servers[s.server_id] = s
+
+    logger.info("Verify tablet replicas distribution")
+    expected_replicas_per_server = get_expected_replicas_per_server(list(live_servers.values()), initial_tablets, rf)
```

Comment:
These two lines are too long & verbose, it took me a minute to understand what is happening. It is also used a few times here (5times in the new tests), so I think it deserves yet another util function. Not sure how to name it:
`expected_replicas_per_server = get_expected_replicas_per_servers(live_servers, dead_servers, initial_tablets, rf)`

Reply:
I can just extend `get_expected_replicas_per_server` and the pre-decommission case can pass an empty list for dead_servers

Reply:
Good idea